hazelcast/src/main/java/com/hazelcast/internal/probing/ProbingCycle.java

```diff
@@ -0,0 +1,222 @@
+ * {@link ProbeSource}s. The {@link ProbeSource} uses the API to communicate its
+ * probes and their current values. 
+ * 
+ * This has two steps:
+ * <ol>
+ * <li>{@link #openContext()} and {@link Tags#tag(CharSequence, CharSequence)}
+ * is used to describe the general context of the probes. The context described
+ * before acts as a "prefix" for all subsequent measurements.</li>
+ * 
+ * <li>{@link #probe(CharSequence, long)} and its sibling methods are used to
+ * state one measurement at a time with its name and value. Alternatively to
+ * stating measurements directly an instance of a type annotated with
+ * {@link Probe} can be passed to {@link #probe(Object)} to measure all
+ * annotated fields and methods.</li>
+ * </ol>
+ * 
+ * Some sources might describe several of such contexts and their probes by
+ * repeating the two steps:
+ * 
+ * <pre>
+ * public void probeIn(ProbingCycle cycle) {
+ *     cycle.openContext().tag(TAG_TYPE, "x").tag(TAG_INSTANCE, "foo");
+ *     cycle.probe(foo);
+ *     cycle.openContext().tag(TAG_INSTANCE, "bar");
+ *     cycle.probe(bar);
+ * }
+ * </pre>
+ */
+@Beta
+@PrivateApi
+public interface ProbingCycle {
+
+    /*
+     * Probing values:
+     */
+
+    /**
+     * Allows {@link ProbeSource}s to optimize their implementations by skipping
+     * handling probes for {@link ProbeLevel}s that are not relevant.
+     * 
+     * {@link ProbeSource}s do not have to obligation to perform this check. They
+     * may just probe everything providing the level for each probe. It is the
+     * {@link ProbingCycle}s obligation to make sure only relevant probes are
+     * considered.
+     * 
+     * @param level the level to check
+     * @return true, if the level is relevant for this cycle, else false
+     */
+    boolean isProbed(ProbeLevel level);
+
+    /**
+     * The main way of rendering metrics is to render instances that have methods
+     * and fields annotated with {@link Probe}. The metric name is derived from
+     * field or method name or {@link Probe#name()} (if given). The metric type
+     * (long or double) is derived from the field type or method return type. The
+     * set of supported types with a known mapping is fix and implementation
+     * dependent.
+     * 
+     * The method exits with the same {@link Tags} context it is called with.
+     * 
+     * @param instance a obj with fields or methods annotated with {@link Probe}
+     */
+    void probe(Object instance);
+
+    /**
+     * Default method to call {@link #probe(Object)} for an array of instances. This
+     * is mostly useful for types that implement {@link Tagging} and thereby provide
+     * their own context.
+     * 
+     * The method exits with the same {@link Tags} context it is called with.
+     * Instances do not affect each others context.
+     * 
+     * @param instances may be null (no effect)
+     */
+    void probe(Object[] instances);
```

Comment:
Can be dangerous call if people create an array every time this method is called.